tools/nimdigger.nim

```diff
@@ -0,0 +1,327 @@
+
+proc runCmdOutput(cmd: string, dir = ""): string =
+  if verbose: dbg cmd, dir
+  let (outp, status) = execCmdEx(cmd, workingDir = dir)
+  doAssert status == 0, indent(&"status: {status}\ncmd: {cmd}\ndir: {dir}\noutput: {outp}", 2)
+  result = outp
+  stripLineEnd(result)
+
+macro ctor(obj: untyped, a: varargs[untyped]): untyped =
+  ## Generates an object constructor call from a list of fields.
+  # xxx expose in std/sugar, factor with https://github.com/nim-lang/fusion/pull/32
+  runnableExamples:
+    type Foo = object
+      a, b: int
+    doAssert Foo.ctor(a,b) == Foo(a: a, b: b)
+  result = nnkObjConstr.newTree(obj)
+  for ai in a: result.add nnkExprColonExpr.newTree(ai, ai)
+
+proc parseKeyVal(a: string): OrderedTable[string, string] =
+  ## parse bash-like entries of the form key=val
+  let a2 = a.splitLines
+  for i, ai in a2:
+    if ai.len == 0 or ai.startsWith "#": continue
+    let b = split(ai, "=", maxsplit = 1)
+    doAssert b.len == 2, $(ai, b)
+    result[b[0]] = b[1]
+
+# xxx move some of these to std/private/gitutils.nim
+proc gitClone(url: string, dir: string) = runCmd fmt"git clone -q {url} {dir.quoteShell}"
+proc gitResetHard(dir: string, rev: string) = runCmd fmt"git -C {dir.quoteShell} reset --hard {rev}"
+proc gitCleanDanger(dir: string, requireConfirmation = true) =
+  #[
+  This is needed to avoid `git bisect` aborting with this error: The following untracked working tree files would be overwritten by checkout.
+  For example, this would happen in cases like this:
+  ```
+  cd $HOME/.nimdigger/cache/Nim
+  git checkout abaa42fd8a239ea62ddb39f6f58c3180137d750c
+  touch testament/testamenthtml.templ
+  cd -
+  nim r tools/nimdigger.nim --oldnew:v0.19.0..v0.20.0 --bisectCmd:'bin/nim -v | grep 0.19.0'
+  ```
+  so we handle cleaning untracked files via dry run (-n) followed by -f if user confirms.
+  ]#
+  let files = runCmdOutput fmt"git -C {dir.quoteShell} clean -n"
+  if files.len > 0:
+    var runClean = true
+    if requireConfirmation:
+      echo &"untracked files may prevent `git bisect` from working, `git -C {dir.quoteShell} clean -n` returned:\n{files}"
+      echo fmt"enter `yes` to proceed with `git clean -f` in: {dir.quoteShell}"
+      let answer = stdin.readLine()
+      runClean = answer == "yes"
+    if runClean:
+      runCmd fmt"git -C {dir.quoteShell} clean -f"
+proc gitFetch(dir: string) = runCmd fmt"git -C {dir.quoteShell} fetch"
+proc gitLatestTag(dir: string): string = runCmdOutput("git describe --abbrev=0 HEAD", dir)
+proc gitCurrentRev(dir: string): string = runCmdOutput("git rev-parse HEAD", dir)
+proc gitCheck(dir: string) =
+  # checks whether we're in a valid git repo; there may be better ways
+  discard runCmdOutput("git describe HEAD", dir)
+
+proc gitIsAncestorOf(dir: string, rev1, rev2: string): bool =
+  gitCheck(dir)
+  execShellCmd(fmt"git -C {dir.quoteShell} merge-base --is-ancestor {rev1} {rev2}") == 0
+
+proc isGitNimTag(tag: string): bool =
```

Comment:
Rather than manual string parsing and comparison, it might be better to use regular expressions or pegs (this applies to the other string comparison logic in the file too).

Reply:
it would be shorter with my own parser library, but doesn't looker simpler/shorter with std/pegs or std/re or nre; feel free to propose if you have some working code that's simpler than what I wrote here though.

Reply:
`strscans` can be used here and can be merged with `parseNimGitTag`

Reply:
thanks, done